File: clojure/binary-search-tree/src/binary_search_tree.clj
(ns binary-search-tree)

(defn singleton [data]
  {:data data})

(def value :data)
(def left :left)
(def right :right)

(defn insert [data node]
  (if (nil? node)
    (singleton data)
    (let [direction (if (<= data (value node)) left right)]
      (assoc node direction (insert data (direction node))))))

(defn to-list [node]
  (if (nil? node)
    '()
    (concat (to-list (left node)) (list (value node)) (to-list (right node)))))

(defn from-list [coll]
  (reduce #(insert %2 %1) nil coll))

File: javascript/binary-search-tree/binary-search-tree.js
export class BinarySearchTree {
  constructor(data) {
    this.data = data;
    this.left = null;
    this.right = null;
  }

  insert(data) {
    if (data <= this.data) {
      if (this.left === null) {
        this.left = new BinarySearchTree(data);
      } else {
        this.left.insert(data);
      }
    } else {
      if (this.right === null) {
        this.right = new BinarySearchTree(data);
      } else {
        this.right.insert(data);
      }
    }
  }

  each(callback) {
    if (this.left !== null) {
      this.left.each(callback);
    }
    callback(this.data);
    if (this.right !== null) {
      this.right.each(callback);
    }
  }
}
